### PR TITLE
Give overview of MIR dataflow framework

### LINKED ISSUE #564: Document MIR dataflow framework
rust-lang/rust#65672 just got merged, so we have a new framework for dataflow analyses in the compiler. I'm in the process of porting analyses that use `BitDenotation` to the new framework (see rust-lang/rust#68241). The new framework has decent in-line documentation, but I think newer contributors would benefit greatly from a high-level overview of dataflow analysis in `rustc`.

### PR DESCRIPTION
**Pull request by ecstatic-morse**
Resolves #564.

I'm a few PRs away from transitioning all existing dataflow analyses in the compiler to the new framework, which can handle gen-kill problems as well as more general ones. When this is finished, we'll want to have an overview of the new framework in the guide.

This is a rough draft, and most of the links don't work since the module structure will change when the old framework is removed. For now, it exists mostly to prove that I've not forgotten about doing this :smile:.

### CODE REVIEW

**Comment by ecstatic-morse** [comment]
I will work on this after some form of rust-lang/rust#71006 lands, since that PR requires changing a lot of the terminology used by the dataflow framework, both in the public API and the implementation.

**Comment by mark-i-m** [comment]
Ping @ecstatic-morse is there a different PR blocking this now?

**Comment by ecstatic-morse** [comment]
Just blocked on me finding the time, unfortunately.

**Comment by mark-i-m** [comment]
Friendly ping @ecstatic-morse :) I know you're busy -- just wanted to make sure this is still on your radar.

**Comment by ecstatic-morse** [comment]
This is not as complete as I would like it to be, but it's much better than nothing and has been updated to include the "arbitrary lattice" extension to the dataflow framework. 

Sorry I didn't complete this in time for @mark-i-m to review it, although I would still welcome their feedback. Maybe @JohnTitor or @jyn514 could take a look?

File: src/mir/dataflow.md
Comment (jyn514) [inline src/mir/dataflow.md:165]:
```
error: Potential incomplete link
   ┌─ mir/dataflow.md:91:63
   │
91 │ Once you have constructed an analysis, you must pass it to an [`Engine`], which
   │                                                               ^^^^^^^^^^ Did you forget to define a URL for ``Engine``?
   │
   = hint: declare the link's URL. For example: `[`Engine`]: http://example.com/`
```
```suggestion
[wiki]: https://en.wikipedia.org/wiki/Data-flow_analysis#Basic_principles
[`Engine`]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_mir/dataflow/struct.Engine.html
```

**Review by jyn514 — APPROVED** [review]
Looks great to me!

**Comment by ecstatic-morse** [comment]
Fixed the broken link and did some line-editing for clarity. Should be good to go now.

**Comment by jyn514** [comment]
Sounds good to me, further improvements can go in follow-ups :)

Thanks so much for writing this up!

**Pull request merged**
